File: tests/main.spec.ts
import { test, expect } from '@playwright/test';

test('navigate redirect and check status codes', async ({ page }) => {
  await page.goto('https://the-internet.herokuapp.com/');
  const title = await page.title();
  test.info().console.log(`title: ${title}`);

  await page.click('text=Redirect Link');
  await page.waitForSelector('#redirect', { state: 'visible' });
  let redirectorHeading: string | null = null;
  try {
    redirectorHeading = (await page.textContent('h3')) || null;
  } catch (e) {
    redirectorHeading = null;
  }
  test.info().console.log(`redirector_heading: ${redirectorHeading}`);

  await page.click('#redirect');
  await page.waitForSelector('div#content ul li a');
  const codes = await page.$$eval('div#content ul li a', els => els.map(e => (e.textContent || '').trim()));
  test.info().console.log(`codes: ${codes.join(',')}`);

  await page.click('text=500');
  await page.waitForSelector('div#content p');
  const contentText = (await page.textContent('#content')) || '';
  test.info().console.log(`content length: ${contentText.length}`);

  expect(contentText).toContain('Status Codes');
});

File: tests/main.spec.py
from playwright.sync_api import sync_playwright


def test_navigate_redirect_and_check_status_codes():
    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True)
        page = browser.new_page()

        page.goto("https://the-internet.herokuapp.com/")
        # Click the Redirect Link
        page.click("text=Redirect Link")

        # Wait for the redirect element with id 'redirect' and click it
        page.wait_for_selector("#redirect", timeout=10000)
        try:
            heading = page.query_selector("h3")
            redirector_heading = heading.inner_text() if heading else None
        except Exception:
            redirector_heading = None
        page.click("#redirect")

        # Collect status codes
        page.wait_for_selector("div#content ul li a", timeout=10000)
        codes = [el.inner_text().strip() for el in page.query_selector_all("div#content ul li a")]

        # Click 500 and verify content
        page.click("text=500")
        page.wait_for_selector("div#content p", timeout=10000)
        content_text = page.query_selector("div#content").inner_text()

        assert "Status Codes" in content_text, f"expected 'Status Codes' in content, got: {content_text!r}"

